### PR TITLE
Add modules to manage Remote Desktop Services

### PR DESCRIPTION
**Pull request by ksubileau**
##### SUMMARY
Added new modules to manage Remote Desktop Services on Windows.

These modules provide a way to setup and configure a part of the Remote Desktop Services, and mainly to build a basic Remote Desktop Gateway (RDG).

- `win_rds_settings` : Configure global Remote Desktop Gateway properties, accessible in the Properties panel of the RD Gateway Manager, and especially the server certificate required to have a working RDG.
- `win_rds_cap` : Create, edit or delete a RDG Connection Authorization Policy (CAP), which manages who can access an associated RDG.
- `win_rds_rap` : Create, edit or delete a RDG Resource Authorization Policy (RAP), which determines which devices a CAP user can connect to via an associated RDG.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
win_rds_cap
win_rds_rap
win_rds_settings

##### ANSIBLE VERSION
```
devel
```

##### ADDITIONAL INFORMATION
Basic usage example:
```yml
- name: install Remote Desktop Services
  win_feature:
    name: RDS-Gateway,RDS-Licensing,RDS-RD-Server
    state: present
    include_management_tools: True

- name: set server certificate for RDS
  win_rds_settings:
    certificate_hash: AFA49C0A75AE798B275FC9FA3466F580DF76B44A

- name: add a simple CAP allowing all users to connect to the RDG
  win_rds_cap:
    name: Simple CAP
    user_groups:
      - users@builtin
    state: enabled

- name: add a simple RAP allowing users to connect to any resource
  win_rds_rap:
    name: Simple RAP
    user_groups:
      - users@builtin
    computer_group_type: any
    state: enabled
```

### CODE REVIEW

**Comment by ksubileau** [comment]
This is still a first draft, especially for the UI, as I'm not sure of the modules and parameters naming and for which parameters should I put a default value. Maybe I also have to improve error handling and test some possible edge cases. 
Also I would like to add support for group names in NetBIOS format (DOMAIN\Username), in addition to UPN format, and maybe handling translations of built-in group names. 
In addition, I think that a fourth module (maybe `win_rds_managed_group`) could be interesting to manage RD Gateway-managed computer groups, used by the module `win_rds_rap`.

Nevertheless, I look forward to read your first feedbacks :)

**Comment by dagwieers** [comment]
So we always expected one PR per module. Not sure if that rule has changed in the meantime.

**Comment by ksubileau** [comment]
Sorry, I didnt't know this rule before the bot reported it to me. As these are 3 modules intended to be used together, and which could possibly be merged or redefined during the development if necessary, it seemed to me easier to group them in a single PR. 
But alright, I can split this PR in 3, one for each module ;)

**Comment by jborean93** [comment]
@ksubileau if they are related let's keep them together for now. I'm trying to make some time to review these but I've been caught up with a few different things. I'll hopefully be able to look at these sometime soon but not sure when sorry.

**Comment by dagwieers** [comment]
Beware that [the deadline](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_7.rst#release-schedule) for getting new features/modules accepted in Ansible v2.7 is nearing, it is set to either 2018-08-23 or 2018-08-30. If you are blocked, or you need feedback, please discuss on IRC channel #ansible-windows or add a comment to the [Windows Working Group meeting agenda](https://github.com/ansible/community/issues/294) to get it resolved.

**Comment by ksubileau** [comment]
I completed the work items I had on my TODO list for this PR, so this is no more a WIP for me.

File: lib/ansible/modules/windows/win_rds_cap.ps1 [hunk 1]
```diff
@@ -0,0 +1,323 @@
+#!powershell
+
+# Copyright: (c) 2018, Ansible Project
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_rds_cap.ps1]: This basically means you are transferring the copyright to the Ansible project ?
Which is fine, but I'd like to make sure that's what you intended.

*(Same for the other modules)*

File: lib/ansible/modules/windows/win_rds_cap.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+description:
+  - Creates, removes and configures a Remote Desktop connection authorization policy (RD CAP).
+  - A RD CAP allows you to specify the users who can connect to a Remote Desktop Gateway server.
+version_added: "2.7"
+author:
+  - Kevin Subileau (@ksubileau)
+options:
+  name:
+    description:
+      - Name of the connection authorization policy
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_rds_cap.py]: All descriptions (except the short_description) have a trailing dot.

*(Same for the other modules)*

File: lib/ansible/modules/windows/win_rds_cap.ps1 [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        [Parameter()]
+        $resultobj = @{}
+    )
+
+    $cap_path = "RDS:\GatewayServer\CAP\$name"
+
+    try {
+        Set-Item -Path "$cap_path\$property" -Value $value -ErrorAction Stop
+    } catch {
+        Fail-Json -obj $resultobj -message "Failed to set property $property of CAP ${name}: $($_.Exception.Message)"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_rds_cap.ps1:93]: I don't think you need to provide a $resultobj to Fail-Json if the content is going to be empty anyway.

*(Same for the other modules)*

File: lib/ansible/modules/windows/win_rds_cap.ps1 [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+        if ($null -ne $session_timeout -and ($session_timeout -ne $cap.SessionTimeout -or $session_timeout_action -ne $cap.SessionTimeoutAction)) {
+            try {
+                Set-Item -Path "RDS:\GatewayServer\CAP\$name\SessionTimeout" `
+                    -Value $session_timeout `
+                    -SessionTimeoutAction ([array]::IndexOf($session_timeout_actions_set, $session_timeout_action)) `
+                    -ErrorAction Stop `
+                    -WhatIf:$check_mode
+            } catch {
+                Fail-Json -obj $resultobj -message "Failed to set property ComputerGroupType of RAP ${name}: $($_.Exception.Message)"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_rds_cap.ps1]: Shouldn't this be `-obj $result` ? I don't think $resultobj is defined in this context.

*(Same for the other modules)*

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_rds_cap.ps1 (hunk 1 above)
Comment (ksubileau) [reply lib/ansible/modules/windows/win_rds_cap.ps1]: This is not very important to me, but as there is also an inconsistency in the file win_rds_settings.py, I will correct.

File: lib/ansible/modules/windows/win_rds_cap.py (hunk 2 above)
Comment (ksubileau) [reply lib/ansible/modules/windows/win_rds_cap.py]: Fixed

File: lib/ansible/modules/windows/win_rds_cap.ps1 (hunk 3 above)
Comment (ksubileau) [reply lib/ansible/modules/windows/win_rds_cap.ps1:93]: I think it costs nothing to leave it. If a future change add something to $result before this line, we will not need to remember that we have to change this line

File: lib/ansible/modules/windows/win_rds_cap.ps1 (hunk 4 above)
Comment (ksubileau) [reply lib/ansible/modules/windows/win_rds_cap.ps1]: Well spotted, fixed

File: lib/ansible/modules/windows/win_rds_cap.ps1 (hunk 3 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_rds_cap.ps1:93]: True.

**Review by dagwieers — APPROVED** [review]
Have not tested the modules myself, but I don't see anything to comment on.

**Comment by ksubileau** [comment]
Here are some points I'm not sure and that could be discussed:

- Modules prefix: win_rd**s** (for Remote Desktop **Services**) or win_rd**g** (for Remote Desktop **Gateway**) ?
- Module naming: `win_rds_cap` and `win_rds_rap` differs only by one letter. That may be confusing or prone to errors. But `win_rds_connection_authorization_policy` is very long... Maybe `win_rds_connection_policy`.
- Which values should return the module ? Is it interesting to return an object representing the CAP/RAP after changes ?
- In module `win_rds_cap`, it may be possible to find a better UI for the device redirection parameters, instead of repeating `redirect_clipboard`, `redirect_drives`, `redirect_...`
- In module `win_rds_cap`, the name for the parameter `allow_only_sdrts_servers` may be improved (quite long and maybe not really clear)
- In module `win_rds_rap`, the choices for `computer_group_type` may be improved (especially `ad_network_resource_group`)
- In module `win_rds_settings`, for the parameter `max_connections`, it may be better to find another value to say "unlimited" instead of the current `-1`
- In module `win_rds_settings`, the name of the parameter `enable_only_messaging_capable_clients` is quite long, maybe it can be shorten

Also I have a general question about Ansible behavior: when a user doesn't use a module parameter, does this mean that the module should not do any change to the corresponding parameter, or that it must set a default value?

**Comment by dagwieers** [comment]
In general I think it's better to stick to the internal/user-facing names, even if they are longer.

> Also I have a general question about Ansible behavior: when a user doesn't use a module parameter, does this mean that the module should not do any change to the corresponding parameter, or that it must set a default value?

Yes, it should not do any changes for parameters that were not set. Parameters not specified default to $null for this reason. In some cases (i.e. creating new objects) default values need to be provided (the first time) and those can be specified in the parameter description if this is the case. (But these are not considered default values, as those would be set everytime a parameter was not specified)

**Comment by ksubileau** [comment]
Ok perfect, that's what i guessed (even if some modules don't seem to fully respect this, as for example... win_nssm!) and implemented, like here:
https://github.com/ansible/ansible/blob/77753a2b18d9b7f96f393e0d0c0b0ad4687dbd61/lib/ansible/modules/windows/win_rds_rap.ps1#L153-L156

File: lib/ansible/modules/windows/win_rds_cap.ps1 [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+
+function Get-CAP([string] $name) {
+    $cap_path = "RDS:\GatewayServer\CAP\$name"
+    $cap = @{
+        Name = $name
+    }
+
+    # Fetch CAP properties
+    Get-ChildItem -Path "$cap_path" | foreach { $cap.Add($_.Name,$_.CurrentValue) }
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_cap.ps1]: Just an FYI, you don't need to quote a variable when used in a parameter value. Quoting is only necessary if you use a literal string and that has a space.

File: lib/ansible/modules/windows/win_rds_cap.ps1 [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+# Validate user groups
+if ($null -ne $user_groups) {
+    if ($user_groups.Count -lt 1) {
+        Fail-Json -obj $result -message "Parameter 'user_groups' cannot be an empty list."
+    }
+
+    $user_groups = $user_groups | foreach {
+        $group = $_
+        # Test that the group is resolvable on the local machine
+        $sid = Convert-ToSID -account_name $group
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_cap.ps1:116]: This will actually call `Fail-Json` if it is unable to resolve the SID. Did you come across a situation where this didn't happen?

File: lib/ansible/modules/windows/win_rds_cap.ps1 [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        if (!$sid) {
+            Fail-Json -obj $result -message "$group is not a valid user group on the host machine or domain"
+        }
+
+        # Return the normalized group name in UPN format
+        $group_name = Convert-FromSID -sid $sid
+        ($group_name -split "\\")[1..0] -join "@"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_cap.ps1]: This doesn't seem right, e.g. `DOMAIN\User` would turn into `User@DOMAIN` when the domain part could really be `DOMAIN.ORG` or something like that. You might need cast the `$sid` string to a SecurityIdentifier object and find a better way to get the UPN from there. Why are you converting to the UPN in the first place?

If you are just using it to get a comparison string then you should be comparing the SIDs themselves.

File: lib/ansible/modules/windows/win_rds_cap.ps1 [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+                $result.changed = $true
+            }
+
+            if($computer_groups_diff) {
+                $diff_text += "~ComputerGroups`n$computer_groups_diff"
+            }
+        }
+
+        # ASK Should we return the full updated CAP object even if the Get-CAP function is a little bit slow ?
+        # $result.cap = Get-CAP -Name $name
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_cap.ps1]: No, use a facts module to do something like this.

File: lib/ansible/modules/windows/win_rds_cap.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: win_rds_cap
+short_description: Manage Connection Authorization Policies (CAP) on a Remote Desktop Gateway server
+description:
+  - Creates, removes and configures a Remote Desktop connection authorization policy (RD CAP).
+  - A RD CAP allows you to specify the users who can connect to a Remote Desktop Gateway server.
+version_added: "2.7"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_cap.py]: Need to bump to 2.8.

File: lib/ansible/modules/windows/win_rds_cap.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+  redirect_serial:
+    description:
+      - Allow serial port redirection.
+    type: bool
+  redirect_pnp:
+    description:
+      - Allow Plug and Play devices redirection.
+    type: bool
+requirements:
+  - Windows Server 2008R2 (6.1) or higher with RDS features
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_cap.py]: Would be good to specify the name of the feature as shown in `Get-WindowsFeature`

File: lib/ansible/modules/windows/win_rds_cap.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+'''
+
+EXAMPLES = '''
+- name: Create a new RDS CAP with a 30 minutes timeout and clipboard redirection enabled
+  win_rds_cap:
+    name: My CAP
+    user_groups:
+      - BUILTIN\\users
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_cap.py]: If you enclose this in `EXAMPLES = r'''` you don't need to escape the `\`. Standard practice for all module doc blocks is to use `r'''  '''`

File: lib/ansible/modules/windows/win_rds_rap.ps1 [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    }
+
+    # Fetch RAP user groups
+    $rap.UserGroups = @(Get-ChildItem -Path "$rap_path\UserGroups" | Select -ExpandProperty Name)
+
+    return $rap
+}
+
+function Set-RAPPropertyValue {
+    [CmdletBinding(SupportsShouldProcess=$true)]
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_rap.ps1:66]: We've gotten rid of this linting rule as Ansible modules do things slightly differently from normal PowerShell cmdlets. Better to get rid of this as it technically doesn't support this.

File: lib/ansible/modules/windows/win_rds_rap.ps1 [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+    }
+    # Ensure the group name is in UPN format
+    $computer_group = Convert-FromSID -sid $sid
+    $computer_group = ($computer_group -split "\\")[1..0] -join "@"
+}
+
+# Validate port numbers
+if ($null -ne $allowed_ports) {
+    foreach ($port in $allowed_ports) {
+        if ($port -notmatch "^([1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5]|any)$") {
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_rap.ps1]: Regex is a bit strange and confusing to me, can you just verify if the int falls within a range of numbers instead?

File: lib/ansible/modules/windows/win_rds_rap.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+DOCUMENTATION = '''
+---
+module: win_rds_rap
+short_description: Manage Resource Authorization Policies (RAP) on a Remote Desktop Gateway server
+description:
+  - Creates, removes and configures a Remote Desktop resource authorization policy (RD RAP).
+  - A RD RAP allows you to specify the network resources (computers) that users can connect
+    to remotely through a Remote Desktop Gateway server.
+version_added: "2.7"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_rap.py]: Bump to 2.8.

File: lib/ansible/modules/windows/win_rds_rap.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+      - 'C(rdg_group): RD Gateway-managed group'
+      - 'C(ad_network_resource_group): Active Directory Domain Services network resource group'
+      - 'C(allow_any): Allow users to connect to any network resource.'
+    choices: [ rdg_group, ad_network_resource_group, allow_any ]
+  computer_group:
+    description:
+      - The computer group name that is associated with this resource authorization policy (RAP).
+      - This is required when I(computer_group_type) is C(rdg_group) or C(ad_network_resource_group).
+requirements:
+  - Windows Server 2008R2 (6.1) or higher with RDS features
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_rap.py]: Put the name of the feature as shown by `Get-WindowsFeature`

File: lib/ansible/modules/windows/win_rds_rap.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+'''
+
+EXAMPLES = '''
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_rap.py]: escape with `r'''` and you don't need to escape `\` in the string itself.

File: lib/ansible/modules/windows/win_rds_settings.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: win_rds_settings
+short_description: Manage main settings of a Remote Desktop Gateway server
+description:
+  - Configure general settings of a Remote Desktop Gateway server.
+version_added: "2.7"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_rds_settings.py]: 2.8 bump

File: test/sanity/pslint/ignore.txt [hunk 18]
```diff
@@ -83,6 +83,8 @@ lib/ansible/modules/windows/win_power_plan.ps1 PSUseDeclaredVarsMoreThanAssignme
 lib/ansible/modules/windows/win_psmodule.ps1 PSAvoidUsingCmdletAliases
 lib/ansible/modules/windows/win_rabbitmq_plugin.ps1 PSAvoidUsingCmdletAliases
 lib/ansible/modules/windows/win_rabbitmq_plugin.ps1 PSAvoidUsingInvokeExpression
+lib/ansible/modules/windows/win_rds_cap.ps1 PSAvoidUsingCmdletAliases
```
Comment (jborean93) [inline test/sanity/pslint/ignore.txt]: We would prefer is you followed this rule and not ignore it.

**Review by jborean93 — COMMENTED** [review]
A few questions but overall looks great.

File: lib/ansible/modules/windows/win_rds_cap.ps1 (hunk 6 above)
Comment (ksubileau) [reply lib/ansible/modules/windows/win_rds_cap.ps1:116]: No I didn't, it's just a bad copy/paste from other modules, like [win_owner#L25-L28](https://github.com/ansible/ansible/blob/277bb3066c7ae8506a8201f5285b77d51b422267/lib/ansible/modules/windows/win_owner.ps1#L25-L28) and [win_share#L26-L29](https://github.com/ansible/ansible/blob/277bb3066c7ae8506a8201f5285b77d51b422267/lib/ansible/modules/windows/win_share.ps1#L26-L29), where I took inspiration.
Also [the error message verified in the tests](https://github.com/ansible/ansible/blob/77753a2b18d9b7f96f393e0d0c0b0ad4687dbd61/test/integration/targets/win_rds_cap/tasks/tests.yml#L254) is the one from the cmdlet, and not the one in the module.

That said, I think it might be a little better to leave the `Fail-Json` call in the module side in this case, to let the calling code throw the error with a more specific message, depending on the context. Currently the message will be `account_name [...] is not a valid account`, while we are actually dealing with groups. The message `[...] is not a valid computer group [...]` is a little bit more precise here.
That's not a big deal, but maybe we should add a silent flag to `Convert-ToSID` to allow the calling code to handle errors (or even ignore it), or reword the embedded error message, or something else... ?

File: lib/ansible/modules/windows/win_rds_cap.ps1 (hunk 7 above)
Comment (ksubileau) [reply lib/ansible/modules/windows/win_rds_cap.ps1]: No it's not for comparison, the UPN form is required for the RDS PSProvider:
```powershell
# This doesn't seem to work if $group is not in UPN form
New-Item -Path "RDS:\GatewayServer\CAP\Test_CAP\UserGroups" -Name $group
```
And that gets on my nerve, because I know this is a quick and dirty conversion but I didn't really find a better way to do it. [Through .NET, there is the DirectoryEntry object](https://stackoverflow.com/a/7524210) but it seems to rely on directory services, if I understand correctly.
If we don't find a better way, maybe we could either use WMI instead of PSProvider, or say that the UPN form is the one recommended for the parameter, and the others are supported for simple cases only...

Comment (jborean93) [reply lib/ansible/modules/windows/win_rds_cap.ps1]: Let’s find the proper API to configure this and not really on editing the registry directly. If it requires DirectoryServices then that’s not the end of the world, it’s just an Add-Type cmdlet away.

File: lib/ansible/modules/windows/win_rds_rap.ps1 (hunk 12 above)
Comment (ksubileau) [reply lib/ansible/modules/windows/win_rds_rap.ps1:66]: That's mean it's better to use a custom argument (like `-CheckMode`) or the global variable `$check_mode` instead of implementing the standard `-WhatIf` argument, is that right ?
The latter has the advantage of being automatically propagated in the call stack.

File: lib/ansible/modules/windows/win_rds_cap.ps1 (hunk 6 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_rds_cap.ps1:116]: Interesting it must be throwing the error earlier than what is being caught in the cmdlet. I agree we should move it out or just have the module until throw and exception instead for finer control in the module.

Comment (ksubileau) [reply lib/ansible/modules/windows/win_rds_cap.ps1:116]: I just saw that there is also one place where the call to Convert-ToSID is surrounded by a try/catch block:
https://github.com/ansible/ansible/blob/277bb3066c7ae8506a8201f5285b77d51b422267/lib/ansible/modules/windows/win_audit_rule.ps1#L60-L66
I don't think there is an unhandled error in the cmdlet that would need to be caught in the calling code.

I suggest that we simply replace the call to `Fail-Json` in `Convert-ToSID` by `return $null`, to let the calling code test for `$null` and throw an error if needed:
```diff
@@ -79,13 +79,13 @@ Function Convert-ToSID {
     try {
         $account_sid = $account.Translate([System.Security.Principal.SecurityIdentifier])
     } catch {
-        Fail-Json @{} "account_name $account_name is not a valid account, cannot get SID: $($_.Exception.Message)"
+        return $null
     }

     return $account_sid.Value
 }
```
Do you agree with that?

Comment (jborean93) [reply lib/ansible/modules/windows/win_rds_cap.ps1:116]: I would investigate this in another issue/pr. My original comment was merely an observation and don’t wish to make a change to a module_util that could affect other modules.

File: lib/ansible/modules/windows/win_rds_cap.ps1 (hunk 7 above)
Comment (ksubileau) [reply lib/ansible/modules/windows/win_rds_cap.ps1]: I havn't been able to find a simple and reliable way to get the group UPN from its SID, working for both local groups and domain groups, so I decide to use WMI methods instead.

**Comment by ksubileau** [comment]
Is there still any corrections to do or is it ready for merging?

**Comment by ksubileau** [comment]
ready_for_review

**Comment by dagwieers** [comment]
During the [next Windows WG meeting](https://github.com/ansible/community/issues/420) on 2019-01-29 we will review and decide on the inclusion of this module in Ansible v2.8. Your presence in the meeting could be helpful.

*If there's anything you wanted to improve/fix, please do so before the meeting.*

**Comment by jborean93** [comment]
Thanks @ksubileau for your work on this one, sorry for taking so long to get back to it. The tests are passing CI and we are happy to merge this in now.

**Comment by ksubileau** [comment]
Thanks @jborean93 and @dagwieers for your help and your review on this, glad to see it merged now :)
